rectools/visuals/visual_app.py

```diff
@@ -122,42 +173,182 @@ def _group_interactions(
-        for recos_df in self.recos.values():
-            request_ids.update(set(recos_df[Columns.TargetItem].unique()))
+        for reco_df in reco.values():
+            request_ids.update(set(reco_df[Columns.TargetItem].unique()))
         interactions = pd.DataFrame({Columns.TargetItem: list(request_ids), Columns.Item: list(request_ids)})
         return interactions
 
+    @classmethod
+    def _df_to_tables_dict(cls, df: pd.DataFrame, key_col: str) -> TablesDict:
+        res = {}
+        for key, grouped_df in df.groupby(key_col):
+            res[key] = grouped_df.drop(columns=[key_col]).reset_index(drop=True)
+        return res
+
+    @classmethod
+    def _create_requests_df(cls, selected_requests: tp.Dict[tp.Hashable, ExternalId]) -> pd.DataFrame:
+        df = pd.DataFrame(
+            {
+                "request_name": list(selected_requests.keys()),
+                "request_id": list(selected_requests.values()),
+            }
+        )
+        return df
+
+    def save(self, folder_name: str, overwrite: bool = False) -> None:
+        """Save stored data for `VisualApp` widgets. This method is not meant to be used
+        directly. Use `VisualApp` or `ItemToItemVisualApp` class methods instead.
+
+        Parameters
+        ----------
+        folder_name : str
+            Destination folder for data.
+        overwrite : bool, default ``False``
+            Allow to overwrite in the folder files if they already exist.
+        """
+        interactions_df = self._ungroup_interactions(
+            grouped_interactions=self.grouped_interactions, selected_requests=self.selected_requests, id_col=self.id_col
+        )
+        reco_df = self._ungroup_reco(
+            grouped_reco=self.grouped_reco, selected_requests=self.selected_requests, id_col=self.id_col
+        )
+        requests_df = self._create_requests_df(self.selected_requests)
+
+        Path(folder_name).mkdir(parents=True, exist_ok=True)
+        mode = "w" if overwrite else "x"
+        interactions_df.to_csv(Path(folder_name, StorageFiles.Interactions), index=False, mode=mode)
+        reco_df.to_csv(Path(folder_name, StorageFiles.Recommendations), index=False, mode=mode)
+        requests_df.to_csv(Path(folder_name, StorageFiles.Requests), index=False, mode=mode)
+
+    @classmethod
+    def load(cls, folder_name: str) -> "AppDataStorage":
+        r"""Load prepared data for VisualApp widgets. This method is not meant to be used
+        directly. Use `VisualApp` or `ItemToItemVisualApp` class methods instead.
+
+        Parameters
+        ----------
+        folder_name : str
+            Folder where data was saved earlier.
+
+        Returns
+        -------
+        AppDataStorage
+            Data storage class for visualisation widgets.
+        """
+        interactions = pd.read_csv(Path(folder_name, StorageFiles.Interactions))
+        reco = pd.read_csv(Path(folder_name, StorageFiles.Recommendations))
+        selected_requests = dict(
```

Comment:
1. It was strange. But `black` actually didn't allow to keep it on single line
2. I changed code logic to make it more clear. Values could be converted to dict because it was an np.array with shape (-1, 2)